README.md
#xHelp
[![Software License](https://img.shields.io/badge/license-GPL-brightgreen.svg?style=flat)](LICENSE) 

xHelp is designed as a user-friendly helpdesk application for the XOOPS portal system. 


docs/CHANGELOG.txt
Changes in 0.81
- converted to XOOPS 2.5.5 Admin GUI (mamba)

Changes in 0.80
1. Bug fix: Added support for 2.4.x when using nocommon option (trabis)
2. Bug fix: Added some php5.3 compatibility (trabis)
3. Feature: Added date range on search form (trabis)
4. Bug fix: Setting default department not working correctly (trabis)

Changes in 0.79
1. Patch for some bugfixes (iconeb)
2. Feature: Reporting module (eric_juden)
3. Feature: Move subnavigation to menu (eric_juden)
4. Feature: Ability for users to add file attachment when responding to a ticket (eric_juden)
5. Bug Fix: View All Ticket pages and Status / State (eric_juden)
6. Bug Fix: Custom Field Edit (ackbarr)
7. Bug Fix: viewing email attachments/enclosures in IE (ackbarr)
8. Bug Fix: Signiture not added to responses (ackbarr)
9. Bug Fix: Status shows closed - close ticket checkbox unchecked (eric_juden)
10. Bug Fix: Update status / Add response (ackbarr)
11. Feature: FAQ/Knowledgebase integration (eric_juden / ackbarr)

Changes in 0.78
1. Bug Fix: Notification form broken in profile.php (ackbarr)
2. Bug Fix: Ticket edit form broken (eric_juden)
3. Bug Fix: Missing query constants in constants.php (eric_juden / ackbarr)
4. Bug Fix: Removed debug lines when adding staff member (eric_juden)
5. Bug Fix: Missing constants for email template folder permissions (eric_juden)
6. Bug Fix: Staff member search says "Log for User" (eric_juden)
7. Bug Fix: Custom fields not available to event handlers during new_ticket event (ackbarr)

Changes in 0.77
1. Bug Fix: Inactive user still receives notifications (ackbarr)
2. Bug Fix: Incorrect module paths in notification emails (XOOPS 2.2) (ackbarr)
3. Bug Fix: Ticket actions block missing change department (ackbarr)
4. Bug Fix: Print ticket page - staff usernames not showing for response (ackbarr)
5. Bug Fix: Anonymous Ticket submission blocked for existing email addresses (ackbarr)
6. Bug Fix: Missing DB Indexes, timeout in upgrade procedures (ackbarr / eric_juden)
7. Bug Fix: Update Status Button is misleading (ackbarr / iconeb)
8. Bug Fix: Staff Member always has "Add Response" permissions (ackbarr)
9. Bug Fix: Fatal Error when assigning ticket to other departement (ackbarr)
10. Feature: Upload multiple files simultaneously (eric_juden)
11. Feature: User ability to submit file attachment for responses (eric_juden)
12. Feature: Staff ticket lists (eric_juden)
13. Bug Fix: View All Ticket pages showing statuses for wrong state (eric_juden)
14. Bug Fix: Numeric custom fields displaying improperly (iconeb / eric_juden)

Changes in 0.75
1. Ability of staff members to merge tickets (eric_juden)
2. Add custom statuses (eric_juden)
3. XOOPS group permissions for logging tickets in specific departments (eric_juden)
4. Allow more than just the ticket submitter be notified via email of ticket changes (eric_juden)
5. Portugues Language Added (_Vlad_)
6. Ability to give reason for changing status (eric_juden)
7. Ability to activate / deactivate mailboxes (ackbarr)
8. Ability to test mail settings prior to saving (ackbarr)
9. Ability to receive notifications via private message (eric_juden)
10. Able to modify an existing department mailbox (eric_juden)
11. Ability to remove file attachments (eric_juden)
12. Ability of staff members to add a response while adding a ticket (eric_juden)
13. Bug Fix: PEAR installed but not in open_basedir path (ackbarr)
14. Administration UI Cleanup (ackbarr)
15. Bug Fix: Take/Assign Ownership Batch Action not working (eric_juden)
16. Bug Fix: Previous response sometimes displayed in Add Response page (eric_juden)
17. Bug Fix: Announcments list from all categories in News 1.2.1 (eric_juden)
18. Bug Fix: xoops_ smarty tags not filled in print ticket page (eric_juden)
19. Bug Fix: Email sent from one dept mailbox to another causes a loop (ackbarr)
20. Bug Fix: Email Attachments with spaces in filename not stored properly (ackbarr)
21. Ability to set number of characters in open tickets block (marcan)
22. Updates to notification service for removing action submitter from notification (marcan)
23. User avatar displays for ticket responses (marcan)
24. XOOPS 2.2 Compatibility (eric_juden)
25. Ability to manage uploaded files from admin side (eric_juden)
26. Bug Fix: Username not being display for staff ratings on Profile page (eric_juden)
27. Ability to display real name instead of username in module preferences (eric_juden)
28. Uploaded files now display the uploaded filename instead of having the ticketid appended to the ticket (eric_juden)
29. End user can now see ticket id on ticket page (eric_juden)
30. Updated Manage Mimetypes page with better searching (eric_juden)
31. Display users in a department on the Edit Department page (eric_juden)
32. Ability to set department visibility by XOOPS group (eric_juden)
33. Ability to save searches (eric_juden)
34. Patch for Manage staff to display grid view of departments and users (iconeb)
35. Ability to add response to ticket upon ticket creation (eric_juden)
36. Ability to search by one or multiple departments (iconeb)
37. Ability to change department from ticket page (iconeb)
38. Ability for staff to go directly to a ticket if they know the ticket id - in staff header (iconeb)
39. Added xhelp main actions block (iconeb)

Changes in 0.71
1. Fixed email activation link not being sent to new users (eric_juden)
2. Fixed uploading files with undefined mimetypes bug (eric_juden)
3. Fixed bug for ticket subject length being too short in edit ticket (eric_juden)
4. Fixed bug for fatal error while in staff profiles (eric_juden)
5. Fixed bug for users being allowed to look at other user's tickets (eric_juden)

Changes in 0.7
1. Auto-create user profiles during ticket submission (eric_juden / ackbarr)
2. Improved Form validation during ticket / response submission (eric_juden)
3. Display available filetypes for upload (eric_juden)
4. Customizable Permission roles for staff members (eric_juden)
5. Email template modifications (alanjuden)

profile.php
<?php
//$Id: profile.php,v 1.42 2005/11/29 17:52:26 ackbarr Exp $
require_once('header.php');
include_once(XHELP_BASE_PATH.'/functions.php');

// Disable module caching in smarty
$xoopsConfig['module_cache'][$xoopsModule->getVar('mid')] = 0;

if($xoopsUser){
    $responseTplID = 0;

    $op = 'default';
    if(isset($_REQUEST['op'])){
        $op = $_REQUEST['op'];
    }

    if(isset($_GET['responseTplID'])){
        $responseTplID = intval($_GET['responseTplID']);
    }

    $xoopsOption['template_main'] = 'xhelp_staff_profile.html';   // Set template
    require(XOOPS_ROOT_PATH.'/header.php');                     // Include the page header

    $numResponses = 0;
    $uid = $xoopsUser->getVar('uid');
    $hStaff =& xhelpGetHandler('staff');
    if (!$staff =& $hStaff->getByUid($uid)) {
        redirect_header(XHELP_BASE_URL."/index.php", 3, _XHELP_ERROR_INV_STAFF);
        exit();
    }
    $hTicketList =& xhelpGetHandler('ticketList');
    $hResponseTpl =& xhelpGetHandler('responseTemplates');
    $crit = new Criteria('uid', $uid);
    $crit->setSort('name');
    $responseTpl =& $hResponseTpl->getObjects($crit);

    foreach($responseTpl as $response){
        $aResponseTpl[] = array('id'=>$response->getVar('id'),
                              'uid'=>$response->getVar('uid'),
                              'name'=>$response->getVar('name'),
                              'response'=>$response->getVar('response'));
    }
    $has_responseTpl = count($responseTpl) > 0;
    unset($responseTpl);

    $displayTpl =& $hResponseTpl->get($responseTplID);

    switch($op){
        case "responseTpl":
            if(isset($_POST['updateResponse'])){
                if(isset($_POST['attachSig'])){
                    $staff->setVar('attachSig', $_POST['attachSig']);
                    if(!$hStaff->insert($staff)){
                        $message = _XHELP_MESSAGE_UPDATE_SIG_ERROR;
                    }
                }
                if($_POST['name'] == '' || $_POST['replyText'] == ''){
                    redirect_header(XHELP_BASE_URL."/profile.php", 3, _XHELP_ERROR_INV_TEMPLATE);
                }
                if($_POST['responseid'] != 0){
                    $updateTpl =& $hResponseTpl->get($_POST['responseid']);
                } else {
                    $updateTpl =& $hResponseTpl->create();
                }
                $updateTpl->setVar('uid', $uid);
                $updateTpl->setVar('name',$_POST['name']);
                $updateTpl->setVar('response',$_POST['replyText']);
                if($hResponseTpl->insert($updateTpl)){
                    $message = _XHELP_MESSAGE_RESPONSE_TPL;
                } else {
                    $message = _XHELP_MESSAGE_RESPONSE_TPL_ERROR;
                }
                redirect_header(XHELP_BASE_URL."/profile.php", 3, $message);
            } else {        // Delete response template
                $hResponseTpl =& xhelpGetHandler('responseTemplates');
                $displayTpl =& $hResponseTpl->get($_POST['tplID']);
                if($hResponseTpl->delete($displayTpl)){
                    $message = _XHELP_MESSAGE_DELETE_RESPONSE_TPL;
                } else {
                    $message = _XHELP_MESSAGE_DELETE_RESPONSE_TPL_ERROR;
                }
                redirect_header(XHELP_BASE_URL."/profile.php", 3, $message);
            }
            break;

        case "updateNotification":
            $notArray = (is_array($_POST['notifications']) ?  $_POST['notifications'] : array(0));
            $notValue = array_sum($notArray);
            $staff->setVar('notify', $notValue);
            if(isset($_POST['email']) && $_POST['email'] <> $staff->getVar('email')){
                $staff->setVar('email', $_POST['email']);
            }
            if(!$hStaff->insert($staff)){
                $message = _XHELP_MESSAGE_UPDATE_EMAIL_ERROR;

            }
            $message = _XHELP_MESSAGE_NOTIFY_UPDATE;
            redirect_header(XHELP_BASE_URL."/profile.php", 3, $message);
            break;

        case "addTicketList":
            if(isset($_POST['savedSearch']) && ($_POST['savedSearch'] != 0)){
                $searchid = intval($_POST['savedSearch']);
                $ticketList =& $hTicketList->create();
                $ticketList->setVar('uid', $xoopsUser->getVar('uid'));
                $ticketList->setVar('searchid', $searchid);
                $ticketList->setVar('weight', $hTicketList->createNewWeight($xoopsUser->getVar('uid')));

                if($hTicketList->insert($ticketList)){
                    header("Location: ".XHELP_BASE_URL."/profile.php");
                } else {
                    redirect_header(XHELP_BASE_URL."/profile.php", 3, _XHELP_MSG_ADD_TICKETLIST_ERR);
                }
            }
            break;

        case "editTicketList":
            if(isset($_REQUEST['id']) && $_REQUEST['id'] != 0){
                $listID = intval($_REQUEST['id']);
            } else {
                redirect_header(XHELP_BASE_URL."/profile.php", 3, _XHELP_MSG_NO_ID);
            }
            break;

        case "deleteTicketList":
            if(isset($_REQUEST['id']) && $_REQUEST['id'] != 0){
                $listID = intval($_REQUEST['id']);
            } else {
                redirect_header(XHELP_BASE_URL."/profile.php", 3, _XHELP_MSG_NO_ID);
            }
            $ticketList =& $hTicketList->get($listID);
            if($hTicketList->delete($ticketList, true)){
                header("Location: ".XHELP_BASE_URL."/profile.php");
            } else {
                redirect_header(XHELP_BASE_URL."/profile.php", 3, _XHELP_MSG_DEL_TICKETLIST_ERR);
            }
            break;

        case "changeListWeight":
            if(isset($_REQUEST['id']) && $_REQUEST['id'] != 0){
                $listID = intval($_REQUEST['id']);
            } else {
                redirect_header(XHELP_BASE_URL."/profile.php", 3, _XHELP_MSG_NO_ID);
            }
            $up = false;
            if(isset($_REQUEST['up'])){
                $up = $_REQUEST['up'];
            }
            $hTicketList->changeWeight($listID, $up);
            header("Location: ".XHELP_BASE_URL."/profile.php");
            break;

        default:
            $xoopsTpl->assign('xhelp_responseTplID', $responseTplID);
            $module_header = '<!--[if gte IE 5.5000]><script src="iepngfix.js" language="JavaScript" type="text/javascript"></script><![endif]-->';
            $xoopsTpl->assign('xhelp_imagePath', XOOPS_URL .'/modules/xhelp/images/');
            $xoopsTpl->assign('xhelp_has_sig', $staff->getVar('attachSig'));
            if(isset($aResponseTpl)){
                $xoopsTpl->assign('xhelp_responseTpl', $aResponseTpl);
            } else {
                $xoopsTpl->assign('xhelp_responseTpl', 0);
            }
            $xoopsTpl->assign('xhelp_hasResponseTpl', (isset($aResponseTpl)) ? count($aResponseTpl) > 0 : 0);
            if(!empty($responseTplID)){
                $xoopsTpl->assign('xhelp_displayTpl_id', $displayTpl->getVar('id'));
                $xoopsTpl->assign('xhelp_displayTpl_name', $displayTpl->getVar('name'));
                $xoopsTpl->assign('xhelp_displayTpl_response', $displayTpl->getVar('response', 'e'));
            } else {
                $xoopsTpl->assign('xhelp_displayTpl_id', 0);
                $xoopsTpl->assign('xhelp_displayTpl_name', '');
                $xoopsTpl->assign('xhelp_displayTpl_response', '');
            }
            $xoopsTpl->assign('xoops_module_header', $module_header);
            $xoopsTpl->assign('xhelp_callsClosed', $staff->getVar('callsClosed'));
            $xoopsTpl->assign('xhelp_numReviews', $staff->getVar('numReviews'));
            $xoopsTpl->assign('xhelp_responseTime', xhelpFormatTime( ($staff->getVar('ticketsResponded') ? $staff->getVar('responseTime') / $staff->getVar('ticketsResponded') : 0)));
            $notify_method = $xoopsUser->getVar('notify_method');
            $xoopsTpl->assign('xhelp_notify_method', ($notify_method == 1) ? _XHELP_NOTIFY_METHOD1 : _XHELP_NOTIFY_METHOD2);

            if(($staff->getVar('rating') == 0) || ($staff->getVar('numReviews') == 0)){
                $xoopsTpl->assign('xhelp_rating', 0);
            } else {
                $xoopsTpl->assign('xhelp_rating', intval($staff->getVar('rating')/$staff->getVar('numReviews')));
            }
            $xoopsTpl->assign('xhelp_uid', $xoopsUser->getVar('uid'));
            $xoopsTpl->assign('xhelp_rating0', _XHELP_RATING0);
            $xoopsTpl->assign('xhelp_rating1', _XHELP_RATING1);
            $xoopsTpl->assign('xhelp_rating2', _XHELP_RATING2);
            $xoopsTpl->assign('xhelp_rating3', _XHELP_RATING3);
            $xoopsTpl->assign('xhelp_rating4', _XHELP_RATING4);
            $xoopsTpl->assign('xhelp_rating5', _XHELP_RATING5);
            $xoopsTpl->assign('xhelp_staff_email', $staff->getVar('email'));
            $xoopsTpl->assign('xhelp_savedSearches', $aSavedSearches);

            $myRoles =& $hStaff->getRoles($xoopsUser->getVar('uid'), true);
            $hNotification =& xhelpGetHandler('notification');
            $settings =& $hNotification->getObjects(null, true);

            $templates =& $xoopsModule->getInfo('_email_tpl');
            $has_notifications = count($templates);

            // Check that notifications are enabled by admin
            $i = 0;
            $staff_enabled = true;
            foreach($templates as $template_id=>$template){
                if($template['category'] == 'dept'){
                    $staff_setting = $settings[$template_id]->getVar('staff_setting');
                    if($staff_setting == 4){
                        $staff_enabled = false;
                    } elseif($staff_setting == 2){
                        $staff_options = $settings[$template_id]->getVar('staff_options');
                        foreach($staff_options as $role){
                            if(array_key_exists($role, $myRoles)){
                                $staff_enabled = true;
                                break;
                            } else {
                                $staff_enabled = false;
                            }
                        }
                    }

                    $deptNotification[] = array('id'=> $template_id,
                                                'name'=>$template['name'],
                                                'category'=>$template['category'],
                                                'template'=>$template['mail_template'],
                                                'subject'=>$template['mail_subject'],
                                                'bitValue'=>(pow(2, $template['bit_value'])),
                                                'title'=>$template['title'],
                                                'caption'=>$template['caption'],
                                                'description'=>$template['description'],
                                                'isChecked'=>($staff->getVar('notify') & pow(2, $template['bit_value'])) > 0,
                                                'staff_setting'=> $staff_enabled);
                }
            }
            if($has_notifications){
                $xoopsTpl->assign('xhelp_deptNotifications', $deptNotification);
            } else {
                $xoopsTpl->assign('xhelp_deptNotifications', 0);
            }

            $hReview  =& xhelpGetHandler('staffReview');
            $hMembers =& xoops_gethandler('member');
            $crit = new Criteria('staffid', $xoopsUser->getVar('uid'));
            $crit->setSort('id');
            $crit->setOrder('DESC');
            $crit->setLimit(5);

            $reviews =& $hReview->getObjects($crit);

            $displayName =& $xoopsModuleConfig['xhelp_displayName'];    // Determines if username or real name is displayed

            foreach ($reviews as $review) {
                $reviewer = $hMembers->getUser($review->getVar('submittedBy'));
                $xoopsTpl->append('xhelp_reviews', array('rating' => $review->getVar('rating'),
                            'ratingdsc' => xhelpGetRating($review->getVar('rating')),
                            'submittedBy' => ($reviewer ? xhelpGetUsername($reviewer, $displayName) : $xoopsConfig['anonymous']),
                            'submittedByUID' => $review->getVar('submittedBy'),
                            'responseid' => $review->getVar('responseid'),
                            'comments' => $review->getVar('comments'),
                            'ticketid' => $review->getVar('ticketid')));
            }
            $xoopsTpl->assign('xhelp_hasReviews', (count($reviews) > 0));

            // Ticket Lists
            $ticketLists =& $hTicketList->getListsByUser($xoopsUser->getVar('uid'));
            $aMySavedSearches = array();
            $mySavedSearches = xhelpGetSavedSearches(array($xoopsUser->getVar('uid'), XHELP_GLOBAL_UID));
            $has_savedSearches = count($aMySavedSearches > 0);
            $ticketListCount = count($ticketLists);
            $aTicketLists = array();
            $aUsedSearches = array();
            $eleNum = 0;
            foreach($ticketLists as $ticketList){
                $weight = $ticketList->getVar('weight');
                $searchid = $ticketList->getVar('searchid');
                $aTicketLists[$ticketList->getVar('id')] = array('id' => $ticketList->getVar('id'),
                                                                 'uid' => $ticketList->getVar('uid'),
                                                                 'searchid' => $searchid,
                                                                 'weight' => $weight,
                                                                 'name' => $mySavedSearches[$ticketList->getVar('searchid')]['name'],
                                                                 'hasWeightUp' => (($eleNum != $ticketListCount - 1) ? true : false),
                                                                 'hasWeightDown' => (($eleNum != 0) ? true : false),
                                                                 'hasEdit' => (($mySavedSearches[$ticketList->getVar('searchid')]['uid'] != -999) ? true : false));
                $eleNum++;
                $aUsedSearches[$searchid] = $searchid;
            }
            unset($ticketLists);

            // Take used searches to get unused searches
            $aSearches = array();
            foreach($mySavedSearches as $savedSearch){
                if(!in_array($savedSearch['id'], $aUsedSearches)){
                    if($savedSearch['id'] != ""){
                        $aSearches[$savedSearch['id']] = $savedSearch;
                    }
                }
            }
            $hasUnusedSearches = count($aSearches) > 0;
            $xoopsTpl->assign('xhelp_ticketLists', $aTicketLists);
            $xoopsTpl->assign('xhelp_hasTicketLists', count($aTicketLists) > 0);
            $xoopsTpl->assign('xhelp_unusedSearches', $aSearches);
            $xoopsTpl->assign('xhelp_hasUnusedSearches', $hasUnusedSearches);
            $xoopsTpl->assign('xhelp_baseURL', XHELP_BASE_URL);
            break;
    }
} else {
    redirect_header(XOOPS_URL .'/user.php', 3);
}

require(XOOPS_ROOT_PATH.'/footer.php');
